src/kbc_modules/cc_kbc/mod.rs

```diff
@@ -94,4 +127,116 @@ impl Kbc {
+        let kbs_uri = self.kbs_uri().to_string();
+        let parsed_kbs_uri =
+            Url::parse(&kbs_uri).map_err(|e| anyhow!("Parse KBS URL Failed: {:?}", e))?;
+        let auth_request = Request::new(self.tee().to_string());
+        let auth_response = self
+            .http_client()
+            .post(format!("{}/auth", kbs_uri))
+            .header("Content-Type", "application/json")
+            .body(serde_json::to_string(&auth_request)?)
+            .send()
+            .await?;
+
+        let mut cookies = auth_response
+            .headers()
+            .get_all(reqwest::header::SET_COOKIE)
+            .iter()
+            .peekable();
+        if cookies.peek().is_some() {
+            self.cookie_store.set_cookies(&mut cookies, &parsed_kbs_uri);
+        }
+
+        let challenge = auth_response.json::<Challenge>().await?;
+        self.nonce = challenge.nonce.clone();
+
+        let attestation_evidence = self.generate_evidence()?;
+        let attest_response = self
+            .http_client()
+            .post(format!("{}/attest", kbs_uri))
+            .header("Content-Type", "application/json")
+            .header(
+                reqwest::header::COOKIE,
+                self.cookie_store
+                    .cookies(&parsed_kbs_uri)
+                    .ok_or_else(|| anyhow!("Authenticated but cookie_store is still empty."))?,
+            )
+            .body(serde_json::to_string(&attestation_evidence)?)
+            .send()
+            .await?;
+        if attest_response.status() != reqwest::StatusCode::OK {
+            // TODO: Parse Error description payload
+            bail!("KBS attest failed")
+        }
+
+        Ok(())
+    }
+
+    async fn request_kbs_resource(&mut self, resource_url: String) -> Result<Response> {
+        let parsed_kbs_uri =
+            Url::parse(self.kbs_uri()).map_err(|e| anyhow!("Parse KBS URL Failed: {:?}", e))?;
+
+        if self.cookie_store.cookies(&parsed_kbs_uri).is_none() {
+            self.establish_kbs_session().await?;
+        }
+
+        let mut res = self
+            .http_client()
+            .get(&resource_url)
+            .header(
+                reqwest::header::COOKIE,
+                self.cookie_store.cookies(&parsed_kbs_uri).ok_or_else(|| {
+                    anyhow!("Established session but cookie_store is still empty.")
+                })?,
+            )
+            .send()
+            .await?;
+
+        if res.status() == reqwest::StatusCode::UNAUTHORIZED {
+            self.establish_kbs_session().await?;
+        }
+        res = self
+            .http_client()
+            .get(&resource_url)
+            .header(
+                reqwest::header::COOKIE,
+                self.cookie_store.cookies(&parsed_kbs_uri).ok_or_else(|| {
+                    anyhow!("Established session again but cookie_store is still empty.")
+                })?,
+            )
+            .send()
+            .await?;
```

Comment:
instead of repeating this part again, I suggest a loop, something like:

```rust
for attempt in 1..3 {
  log!("trying to get resource, attempt {attempt}");
  request resource
  if UNAUTHORIZED {
    establish_kbs_session
    sleep 5 seconds
  } else {
   break
  }
} 
```